### PR TITLE
Add Custom Field Configuration Editor for Shoot Clusters

### PR DESCRIPTION
**Pull request by petersutter**
**What this PR does / why we need it**:
This PR introduces an editor for custom shoot fields. The new feature allows users to configure custom fields directly from the dashboard without needing to use `kubectl`. 
custom-fields-5
custom-fields-6
custom-fields-7

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```feature user
You can now configure custom shoot fields directly from the dashboard using the new editor, eliminating the need to use `kubectl`.
```

### CODE REVIEW

**Comment by petersutter** [comment]
- cancel / save buttons always visible on dialog
- update project on edit dialog close / delete field
- icon in primary colour
- path input full width
- not defined / unknown / - ?

File: frontend/src/views/GShootItemPlaceholder.vue [hunk 1]
```diff
@@ -208,6 +209,9 @@ export default {
 
     provide('activePopoverKey', activePopoverKey)
 
+    const projectItem = computed(() => projectStore.project)
+
+    useProvideProjectItem(projectItem)
```
[the remark this answers is not in the document]
Comment (petersutter) [reply frontend/src/views/GShootItemPlaceholder.vue:214]: ✅

File: frontend/src/views/GShootList.vue [hunk 2]
```diff
@@ -298,10 +301,19 @@ export default {
...
   setup () {
+    const projectStore = useProjectStore()
+
     const activePopoverKey = ref('')
     provide('activePopoverKey', activePopoverKey)
+
+    const projectItem = computed(() => projectStore.project)
+    const {
+      shootCustomFields,
+    } = useProjectShootCustomFields(projectItem)
```
[the remark this answers is not in the document]
Comment (petersutter) [reply frontend/src/views/GShootList.vue:312]: ✅

File: frontend/src/components/GShootCustomFieldsConfiguration.vue [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+} = useProjectItem()
+
+const {
+  setProjectManifest,
+  getShootCustomFieldsPatchData,
+} = useProjectContext()
+
+const actionDialog = ref(null)
+
+const canPatchProject = computed(() => authzStore.canPatchProject)
```
[the remark this answers is not in the document]
Comment (petersutter) [reply frontend/src/components/GShootCustomFieldsConfiguration.vue:59]: ✅

File: frontend/src/composables/useScrollBar.js [hunk 4]
```diff
@@ -0,0 +1,16 @@
+//
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Gardener contributors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import { watch } from 'vue'
+
+export function useScrollBar (cardContentRef) {
```
[the remark this answers is not in the document]
Comment (petersutter) [reply frontend/src/composables/useScrollBar.js:9]: ✅

File: frontend/src/composables/useProjectItem.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    projectAnnotations,
+    getProjectAnnotation,
+    setProjectAnnotation,
+    unsetProjectAnnotation,
+    projectLabels,
+    getProjectLabel,
+    setProjectLabel,
+    unsetProjectLabel,
+    projectCreatedAt,
+    projectCreatedBy,
```
[the remark this answers is not in the document]
Comment (petersutter) [reply frontend/src/composables/useProjectItem.js:37]: ✅

File: frontend/src/composables/useProjectItem.js [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    isNewProject,
+  } = useProjectMetadata(projectItem)
+
+  const {
+    shootCustomFields,
+  } = useProjectShootCustomFields(projectItem)
+
```
[the remark this answers is not in the document]
Comment (petersutter) [reply frontend/src/composables/useProjectItem.js:44]: ✅

File: frontend/src/views/GAdministration.vue [hunk 7]
```diff
@@ -537,272 +545,181 @@ import {
...
+const {
+  shootCustomFields,
+} = useProvideProjectItem(projectItem)
+
+const namespace = computed(() => authzStore.namespace)
+const canPatchProject = computed(() => authzStore.canPatchProject)
+const canManageMembers = computed(() => authzStore.canManageMembers)
+const canDeleteProject = computed(() => authzStore.canDeleteProject)
+const isKubeconfigEnabled = computed(() => kubeconfigStore.isKubeconfigEnabled)
+const projectQuotaStatus = computed(() => quotaStore.projectQuotaStatus)
```
[the remark this answers is not in the document]
Comment (petersutter) [reply frontend/src/views/GAdministration.vue:579]: ✅

File: frontend/src/views/GAdministration.vue [hunk 8]
```diff
@@ -537,272 +545,181 @@ import {
...
+
+const userList = computed(() => {
+  const members = new Set()
+  for (const { username } of memberStore.memberList) {
+    if (!isServiceAccountUsername(username)) {
+      members.add(username)
+    }
+  }
+  if (projectDetails.value.owner) {
+    members.add(projectDetails.value.owner)
```
[the remark this answers is not in the document]
Comment (petersutter) [reply frontend/src/views/GAdministration.vue:589]: ✅

File: frontend/src/components/GManageShootCustomFields.vue [hunk 9]
```diff
@@ -0,0 +1,555 @@
...
+                    variant="underlined"
+                    label="Weight"
+                    type="number"
+                    hint="Order of the column"
+                    persistent-hint
+                    v-bind="props"
+                    :disabled="!editedField.showColumn"
+                  />
+                </template>
+                <span>The standard columns start with weight 100 and continue in 100 increments (200, 300, ..)</span>
```
Comment (grolu) [inline frontend/src/components/GManageShootCustomFields.vue:268]:
```suggestion
                <span>The built-in columns start with a weight of 100, increasing by 100 (200, 300, etc.).</span>
```

Comment (petersutter) [reply frontend/src/components/GManageShootCustomFields.vue]: ✅

File: frontend/src/components/GManageShootCustomFields.vue [hunk 10]
```diff
@@ -503,7 +517,38 @@ async function addField () {
...
+async function updateConfiguration (onSuccess) {
+  loading.value = true
+  try {
+    const { metadata: { name, namespace } } = projectStore.project
+    const mergePatchDocument = {
+      metadata: { name, namespace },
+    }
+    Object.assign(mergePatchDocument, getShootCustomFieldsPatchData())
+
+    await projectStore.patchProject(mergePatchDocument)
```
Comment (holgerkoser) [inline frontend/src/components/GManageShootCustomFields.vue:538]: The patchData function should be a computed property that contains also the name and namespace:
```js
  const shootCustomFieldsPatchData = computed(() => {
    const { namespace, name } = projectItem.metadata
    return {
      metadata: {
        namespace,
        name,
        annotations: {
          'dashboard.gardener.cloud/shootCustomFields': getProjectAnnotation('dashboard.gardener.cloud/shootCustomFields', null),
        },
      },
    }
  })
```

**Review by holgerkoser — APPROVED** [review]
/lgtm

File: frontend/src/components/GShootListRow.vue [hunk 11]
```diff
@@ -144,23 +144,23 @@ SPDX-License-Identifier: Apache-2.0
...
-          <span v-else>{{ cell.value }}</span>
-        </template>
+        <v-tooltip
+          v-if="cell.header.tooltip"
+          location="top"
+        >
+          <template #activator="slotProps">
+            <span
+              v-bind="slotProps.props"
+              :class="classForValue(cell.value)"
```
Comment (grolu) [inline frontend/src/components/GShootListRow.vue:154]:
```suggestion
              :class="{'text-disabled' : !cell.value}"
```

File: frontend/src/components/GShootListRow.vue [hunk 12]
```diff
@@ -144,23 +144,23 @@ SPDX-License-Identifier: Apache-2.0
...
+            >{{ cell.displayValue }}</span>
+          </template>
+          {{ cell.header.tooltip }}
+        </v-tooltip>
         <span
-          v-else-if="cell.header.defaultValue"
-          class="text-grey"
+          v-else-if="cell.displayValue"
+          :class="classForValue(cell.value)"
```
Comment (grolu) [inline frontend/src/components/GShootListRow.vue:161]:
```suggestion
         :class="{'text-disabled' : !cell.value}"
```

File: frontend/src/components/ShootDetails/GCustomFieldsCard.vue [hunk 13]
```diff
@@ -6,54 +6,100 @@ SPDX-License-Identifier: Apache-2.0
...
-export default {
-  props: {
-    customFields: {
-      type: Array,
-    },
-  },
+function classForValue (value) {
+  return {
+    'text-grey': !value,
+  }
```
Comment (grolu) [inline frontend/src/components/ShootDetails/GCustomFieldsCard.vue:103]:
```suggestion
```

File: frontend/src/components/ShootDetails/GCustomFieldsCard.vue [hunk 14]
```diff
@@ -6,54 +6,100 @@ SPDX-License-Identifier: Apache-2.0
...
             <v-tooltip
-              v-if="customField.tooltip"
+              v-if="field.tooltip"
               location="top"
             >
               <template #activator="{ props }">
-                <span v-bind="props">{{ customField.value }}</span>
+                <span
+                  v-bind="props"
+                  :class="classForValue(field.value)"
```
Comment (grolu) [inline frontend/src/components/ShootDetails/GCustomFieldsCard.vue:40]: :class="{'text-disabled' : !field.value}"

File: frontend/src/components/ShootDetails/GCustomFieldsCard.vue [hunk 15]
```diff
@@ -6,54 +6,100 @@ SPDX-License-Identifier: Apache-2.0
...
+                >{{ field.displayValue }}</span>
               </template>
-              {{ customField.tooltip }}
+              {{ field.tooltip }}
             </v-tooltip>
-            <span v-else>{{ customField.value }}</span>
+            <span
+              v-else-if="field.displayValue"
+              :class="classForValue(field.value)"
```
Comment (grolu) [inline frontend/src/components/ShootDetails/GCustomFieldsCard.vue:47]:
```suggestion
             :class="{'text-disabled' : !field.value}"
```

File: frontend/src/components/ShootDetails/GCustomFieldsCard.vue (hunk 14 above)
Comment (grolu) [inline frontend/src/components/ShootDetails/GCustomFieldsCard.vue:40]:
```suggestion
                  :class="{'text-disabled' : !field.value}"
```

**Pull request merged**